### PR TITLE
[#135] Allow specifying the mount path for ConfigMaps

### LINKED ISSUE #135: With operator it is not possible to mount configMap to $JBOSS_HOME/extensions
With the operator, it is not possible to mount configMap to $JBOSS_HOME/extensions.

We need to be able to mount a ConfigMap by configuring the mountPath of this config map.

### PR DESCRIPTION
**Pull request by yersan**
This PR will allow the configuration of ConfigMpas by defining a `MountPath` in addition to the `Name`

The `MountPath` could be a relative or an absolute path. If it is a relative path, then it is treated as relative to JBOSS_HOME. The `MountPath` is optional, and if it is not specified then it defaults to /etc/configmaps/<config map name>

This fixes #135

### CODE REVIEW

File: pkg/apis/wildfly/v1alpha1/wildflyserver_types.go [hunk 1]
```diff
@@ -130,6 +134,26 @@ type WildFlyServerList struct {
...
+func (in *ConfigMapSpec) ResolveMountPath() string {
+	result := path.Join(constants.ConfigMapsDir, in.Name)
+	if in.MountPath != "" {
+		if path.IsAbs(in.MountPath) {
+			result = in.MountPath
+		} else {
+			result = path.Join(constants.JBossHome, in.MountPath)
+		}
+	}
+	return result
```
Comment (jmesnil) [inline pkg/apis/wildfly/v1alpha1/wildflyserver_types.go]: at this place, we could have a sanity check that there is no `:` in the resulting path

File: deploy/crds/wildfly_v1alpha1_wildflyserver_crd.yaml [hunk 2]
```diff
@@ -47,11 +47,19 @@ spec:
               type: string
             configMaps:
               description: ConfigMaps is a list of ConfigMaps in the same namespace
-                as the WildFlyServer object, which shall be mounted into the WildFlyServer
-                Pods. The ConfigMaps are mounted into /etc/configmaps/<configmap-name>.
-              items:
-                type: string
-              minItems: 1
+                as the WildFlyServer object to mount as a volume into the WildFlyServer
```
[the remark this answers is not in the document]
Comment (yersan) [reply deploy/crds/wildfly_v1alpha1_wildflyserver_crd.yaml]: It was edited manually 👎. I updated it using `make codegen` instead

File: doc/apis.adoc [hunk 3]
```diff
@@ -82,6 +82,16 @@ transaction, make sure to specify a `volumeClaimTemplate` that so that the same
...
 
+[[configmapspec]]
+## `ConfigMapSpec`
+
+`ConfigMapSpec` defines a `ConfigMap` name to mount as volume in the container. By specifying `mountPath` it is possible to configure the path where this config map should be mounted.
+[options="header,footer"]
+|=======================
+| Field  | Description |Scheme| Required
+| `name` | Name of the `ConfigMap` to mount as a volume. | string | true
+| `mountPath` | Path within the volume from which the container's volume should be mounted. Must not contain ':'. In case of a relative path, the path is treated as relative to `JBOSS_HOME`. It defaults to `/etc/configmaps/<config map name>` | string |false
```
[the remark this answers is not in the document]
Comment (yersan) [reply doc/apis.adoc]: We are now rejecting the CR creation by imposing a pattern to do not contain the ':', the advantage is the resource is rejected immediately when the user is creating it.

File: doc/user-guide.adoc [hunk 4]
```diff
@@ -148,29 +148,32 @@ $ kubectl create configmap my-config --from-literal=key1=value1 --from-literal=k
...
 key1 key2
-[jboss@quickstart-0 ~]$ cat /etc/configmaps/my-config/key1
+[jboss@quickstart-0 ~]$ cat /etc/my-config-map/key1
 value1
-[jboss@quickstart-0 ~]$ cat /etc/configmaps/my-config/key2
+[jboss@quickstart-0 ~]$ cat /etc/my-config-map/key2
 value2
 ----
 
+The `mountPath` is optional and it could be a relative or an absolute path. If it is a relative path, the path is treated as relative to JBOSS_HOME. If it is not specified by default the config map is mounted under /etc/configmaps/<config map name>.
```
[the remark this answers is not in the document]
Comment (yersan) [reply doc/user-guide.adoc]: Done, added to the user-guide

File: pkg/apis/wildfly/v1alpha1/wildflyserver_types.go [hunk 5]
```diff
@@ -1,8 +1,10 @@
 package v1alpha1
 
 import (
+	"github.com/wildfly/wildfly-operator/pkg/resources/constants"
```
[the remark this answers is not in the document]
Comment (yersan) [reply pkg/apis/wildfly/v1alpha1/wildflyserver_types.go]: Fixed. Removed the need to move `constants` to a different package.

File: pkg/apis/wildfly/v1alpha1/wildflyserver_types.go [hunk 6]
```diff
@@ -130,6 +134,26 @@ type WildFlyServerList struct {
...
 
+// ConfigMapSpec represents a ConfigMap definition with a name and a mount path
+// +k8s:openapi-gen=true
+type ConfigMapSpec struct {
+	Name      string `json:"name"`
+	MountPath string `json:"mountPath,omitempty"`
+}
+
+// ResolveMountPath resolves the final mount path for the current ConfigMapSpec
+func (in *ConfigMapSpec) ResolveMountPath() string {
```
[the remark this answers is not in the document]
Comment (yersan) [reply pkg/apis/wildfly/v1alpha1/wildflyserver_types.go]: Moved to the statefulset.go

File: pkg/apis/wildfly/v1alpha1/zz_generated.deepcopy.go [hunk 7]
```diff
@@ -9,6 +9,22 @@ import (
 	runtime "k8s.io/apimachinery/pkg/runtime"
 )
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
[the remark this answers is not in the document]
Comment (yersan) [reply pkg/apis/wildfly/v1alpha1/zz_generated.deepcopy.go]: It was edited manually 👎. I updated it using make `codegen instead`

File: pkg/apis/wildfly/v1alpha1/wildflyserver_types.go (hunk 1 above)
Comment (yersan) [reply pkg/apis/wildfly/v1alpha1/wildflyserver_types.go]: Now we are using openapi validation at CR level

File: pkg/apis/wildfly/v1alpha1/wildflyserver_types.go [hunk 8]
```diff
@@ -130,6 +134,26 @@ type WildFlyServerList struct {
 	Items           []WildFlyServer `json:"items"`
 }
 
+// ConfigMapSpec represents a ConfigMap definition with a name and a mount path
+// +k8s:openapi-gen=true
+type ConfigMapSpec struct {
+	Name      string `json:"name"`
+	MountPath string `json:"mountPath,omitempty"`
```
[the remark this answers is not in the document]
Comment (yersan) [reply pkg/apis/wildfly/v1alpha1/wildflyserver_types.go]: Do you mean this one? https://github.com/wildfly/wildfly-operator/pull/136/files#diff-553a69f3eee8fa0585c8bfe19a47a871R43-R45

**Comment by pkremens** [comment]
@yersan @jmesnil @mchoma @jbliznak I don't think we can do it that way. This is a way we use to setup the configMaps field in EAP Operator (registry-proxy.engineering.redhat.com/rh-osbs/jboss-eap-7-eap73-rhel8-operator:1.0-8)

```
  configMaps:
  - config-map-name
```

After the change, the same setup would be 

```
    configMaps:
    - name: config-map-name
```

Trying to use the old style config with this version of operator gives me 

```
# * spec.configMaps: Invalid value: "string": spec.configMaps in body must be of type object: "string"
```

thus we're breaking the backward compatibility with this one.

**Comment by jmesnil** [comment]
That's correct, this feature would break backwards compatibility.
Note that the CRD `apiVersion` is `wildfly.org/v1alpha1`. We have clearly identified that our API is in an alpha state and might be changed without backwards compatibility.

We should _strive_ to preserve backwards compatibility when introducing a new feature but in that specific case, I prefer we provide a clean representation of a config map (with a name and an optional location) rather than preserve the backwards compatibility with a clunky api.
When we discussed this feature with @yersan we could not find a correct representation for that feature that would preserve backwards compatibility so I ok-ed to break it.

**Comment by jmesnil** [comment]
let's close this PR as we will need first to have a conversion web hook to introduce this change without breaking compatibility

**Pull request closed without merging**